samcli/commands/init/init_templates.py

```diff
@@ -0,0 +1,150 @@
+        choice = click.prompt("Template Selection", type=click.Choice(choices), show_choices=False)
+        template_md = options[int(choice) - 1]  # zero index
+        if template_md.get("init_location") is not None:
+            return template_md["init_location"]
+        elif template_md.get("directory") is not None:
+            return os.path.join(self.repo_path, template_md["directory"])
+        else:
+            raise UserException("Invalid template. This should not be possible, please raise an issue.")
+
+    def location_from_app_template(self, runtime, dependency_manager, app_template):
+        options = self.init_options(runtime, dependency_manager)
+        try:
+            template = next(item for item in options if self._check_app_template(item, app_template))
+            if template.get("init_location") is not None:
+                return template["init_location"]
+            elif template.get("directory") is not None:
+                return os.path.join(self.repo_path, template["directory"])
+            else:
+                raise UserException("Invalid template. This should not be possible, please raise an issue.")
+        except StopIteration:
+            msg = "Can't find application template " + app_template + " - check valid values in interactive init."
+            raise UserException(msg)
+
+    def _check_app_template(self, entry, app_template):
+        return entry["appTemplate"] == app_template
+
+    def init_options(self, runtime, dependency_manager):
+        if self.clone_attempted is False:
+            self._clone_repo()
+        if self.repo_path is None:
+            return self._init_options_from_bundle(runtime, dependency_manager)
+        return self._init_options_from_manifest(runtime, dependency_manager)
+
+    def _init_options_from_manifest(self, runtime, dependency_manager):
+        manifest_path = os.path.join(self.repo_path, "manifest.json")
+        with open(str(manifest_path)) as fp:
+            body = fp.read()
+            manifest_body = json.loads(body)
+            templates = manifest_body.get(runtime)
+            if templates is None:
+                # Fallback to bundled templates
+                return self._init_options_from_bundle(runtime, dependency_manager)
+            if dependency_manager is not None:
+                templates_by_dep = filter(lambda x: x["dependencyManager"] == dependency_manager, templates)
+                return list(templates_by_dep)
+            return templates
+
+    def _init_options_from_bundle(self, runtime, dependency_manager):
+        for mapping in list(itertools.chain(*(RUNTIME_DEP_TEMPLATE_MAPPING.values()))):
+            if runtime in mapping["runtimes"] or any([r.startswith(runtime) for r in mapping["runtimes"]]):
+                if not dependency_manager or dependency_manager == mapping["dependency_manager"]:
+                    mapping["appTemplate"] = "hello-world"  # when bundled, use this default template name
```

Comment:
do we want the logic of falling back to the bundled be driven by manifest in an external repository? https://github.com/awslabs/aws-sam-cli-app-templates/blob/master/manifest.json#L102

Reply:
I am not sure that's possible. If we get to this piece of code, the manifest is not accessible. Or is this a separate scenario you're thinking about?

Reply:
No, I was just thinking that falling back to bundled templates (within the same repo), required direction from a manifest file in another repo (app templates repo), introduces some coupling between the repos.